Data/HashMap/Base.hs

```diff
@@ -366,9 +377,12 @@ equalKeys1 eq t1 t2 = go (toList' t1 []) (toList' t2 [])
     go (Leaf k1 l1 : tl1) (Leaf k2 l2 : tl2)
       | k1 == k2 && leafEq l1 l2
       = go tl1 tl2
-    go (Collision k1 ary1 : tl1) (Collision k2 ary2 : tl2)
-      | k1 == k2 && A.length ary1 == A.length ary2 &&
-        isPermutationBy leafEq (A.toList ary1) (A.toList ary2)
+    go (Collision h1 (L k11 _) (L k12 _) hm1 : tl1) (Collision h2 (L k21 _) (L k22 _) hm2 : tl2)
+      | h1 == h2 &&
+        (length hm1 == length hm2) &&
+        isPermutationBy eq
+          (k11 : k12 : keys hm1)
+          (k21 : k22 : keys hm2)
```

Comment:
This is most definitely wrong. You're ignoring the values altogether here!

Reply:
Isn't that the point of `equalKeys`, which has the comment "Same as 'equal' but doesn't compare the values"?
Note that the old and the new semantics are the same.

Reply:
No. `equalKeys` is only for seeing if two things have the same keys (e.g., to implement equality testing for hash *sets*). It has nothing to do with what we're doing *here*.

Reply:
However, there's still something to fix here, I think. Can't we add a `Hashable` constraint and do the same optimization as for real equality?

Reply:
@treeowl I can do that for `equalKeys`, but not for `equalKeys1` because we need a `Hashable` constraint, and we don't get that from `Eq1`

Reply:
Right. The `Eq2` and `Ord2` instances for maps and the `Eq1` and `Ord1` instances for sets are hopeless for large collision buckets. That doesn't mean we shouldn't fix what's fixable.

Reply:
> That doesn't mean we shouldn't fix what's fixable.

Sorry, I didn't mean to sound like I was suggesting that.